source/agora/common/Hash.d

```diff
@@ -289,3 +289,182 @@ nothrow @nogc @safe unittest
+        (1000 / 5).
+
+        Params:
+          count = Number of samples to store (number of entries in the array)
+          sample_size = Distance between two pre-images
+
+    ***************************************************************************/
+
+    public this (ulong count, ulong sample_size) pure
+    {
+        assert(count > 1, "A count of less than 2 does not make sense");
+        assert(sample_size > 1, "A distance of less than 2 does not make sense");
+
+        this.interval = sample_size;
+        this.data = new Hash[](count);
+    }
+
+    /***************************************************************************
+
+        Populate this cache from the `seed`
+
+        Params:
+          seed = Initial value to derive preimage from. This will always be the
+                 first value of the array.
+          length = The number of entries to populate.
+                   This can be used when large sample size are used,
+                   and one wish to stop initialization past a certain threshold.
+                   The special value `0` means "everything", and is the default.
+
+        Returns:
+          The result of hashing seed `sample_size * (count + 1) - 1` times
+          if `length == 0`, or `sample_size * (length + 1) - 1` otherwise,
+          provided `length <= count`.
+
+    ***************************************************************************/
+
+    public Hash reset (Hash seed, size_t length = 0)
+    {
+        // Set the length, so that extra entries are not accessible through
+        // `opIndex`
+        if (length)
+        {
+            this.data.length = length;
+            () @trusted { assumeSafeAppend(this.data); }();
+        }
+        size_t count = 0;
+        immutable end = this.data.length * this.interval - 1;
+        foreach (idx, ref entry; this.data)
+        {
+            entry = seed;
+            do
+                seed = hashFull(seed);
+            while (++count % end % this.interval);
+        }
+        return seed;
+    }
+
+    /***************************************************************************
+
+        Get the hash at index `idx` in the cycle
+
+        Will perform at most `cycle_length % interval` computations.
+
+    ***************************************************************************/
+
+    public Hash opIndex (size_t index) const @nogc
+    {
+        immutable startIndex = (index / this.interval);
+        // This will trigger out of bound if needed
+        // We could possibly silently allow to index arbitrarily outside the
+        // array by just taking the computational hit, since indexing past the
+        // end means we are generating new pre-images, but it sounds like abuse
+        Hash value = this.data[startIndex];
+        foreach (_; (startIndex * this.interval) .. index)
+            value = hashFull(value);
+        return value;
+    }
+}
+
+///
```

Comment:
I was just thinking about this. The documented examples would ideally show a use-case. Here they're just elaborate tests that try to do full coverage.

But this is true right now for most of our tests. In the future we might want to provide better _documented_ examples.

Reply:
Oh right. Yeah definitely. Should we open an issue to "curate" the documentation ? TestNet milestone plz.